src/language-js/print/assignment.js

```diff
@@ -284,15 +284,27 @@ function hasComplexTypeAnnotation(node) {
   const typeParams = getTypeParametersFromTypeReference(
     typeAnnotation.typeAnnotation
   );
-  return (
+
+  // Complex type
+  if (
     isNonEmptyArray(typeParams) &&
     typeParams.length > 1 &&
     typeParams.some(
       (param) =>
         isNonEmptyArray(getTypeParametersFromTypeReference(param)) ||
         param.type === "TSConditionalType"
     )
-  );
+  ) {
+    return true
+  }
+
+  // Simple but long type
+  const widthOfStatementBeforeEquals = 1000 // TODO (Halp!)
```

Comment:
It's okay that it's difficult. Assignments are one of the most complex parts of Prettier's JS printer. I can't look into this deeply at the moment, but can give some general advice.

"I should determine if the line is too long" and "place to inject (\nType\n) =" is not really how Prettier works. The intermediate representation (AKA _doc_) that the printer builds doesn't depend on the line length limit.

You can read about the IR [here](https://github.com/prettier/prettier/blob/main/commands.md) and play with it [here](https://deploy-preview-10183--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBzAThArgBwBQDaAOlAASlEh4BUlANCeQJZQAmCMhj5pANi3AZkelAAro4MGEzjok9buX5RBiiiADyOFk2gBDeKwYghPPgNMiQAYQjtjlpRbWUAYhHQBbA-HQLhALoAlJbKqsKU1EH+wSYgEDjS0ADOyKB66JgA7qIZCKkoerxZegCeqXQgAEboemAA1pIAyjh1LKjIMOhYgiBwnlVwrOysADJ6UKhYeqhw7l4+7cggelgwEHEAFjCevADqm0zwya1gcE35R0wAbkely2DJFSAsybIw4jPeyABmRW+VABWyQAHgAhWoNZp6TxwUYCX7-XrAkFNdq8OAARSwEHgiN4AJArXQb3Qy1YEDAAFo4CCcLwPLI4jh0CwYHsmKwYJtkAAOAAMlRZEDee1qOGWLLgpOucDiAEccfBxAkCitklSVEMhnEJIqmBJPqhvkg-gTem9PExOt0LeisUq5aakZUYHoqhyuTykAAmV21JjKVC2TwmvrJACscSwbwAKu6CmbCdcegBJNgcJpgVmJACCbCaMFKGPxbwAvmWgA). `group` is the main basic building block you should understand to make Prettier do what you need.

Reply:
Thanks a lot for this detailed explanation @thorn0! Things make more sense now – I’m back at this task.

@sosukesuzuki & @fisker I see you’ve been making PRs related to TS 4.3 recently, so if I understand correctly a new release should be soon. It’d be great if you could give me a chance to finish this work so that we could include it as well 🙂

If what I’m trying to do looks trivial to you and you know what exactly to change to pass the new tests, that will be awesome! Still not sure how many hours this will take me at this point 😅 It’s now less than infinity, thanks to guidance by @thorn0!

Reply:
> If what I’m trying to do looks trivial

It's not trivial. I guess I should have warned you about that earlier. 🙄